frontend/ui/bidi.lua

```diff
@@ -0,0 +1,157 @@
+--[[--
+Bidirectional text and UI mirroring helpers.
+Some overview at:
+https://material.io/design/usability/bidirectionality.html
+]]
+
+local Language = require("ui/language")
+local _ = require("gettext")
+
+local Bidi = {
+    _mirrored_ui_layout = false,
+    _rtl_ui_text = false,
+}
+
+-- Setup UI mirroring and RTL text for UI language
+function Bidi.setup(lang)
+    local is_rtl = Language:isLanguageRTL(lang)
+    -- Mirror UI if language is RTL
+    Bidi._mirrored_ui_layout = is_rtl
+    -- Unless requested not to (or requested mirroring with LTR language)
+    if G_reader_settings:isTrue("reverse_lang_ui_layout_mirroring") then
+        Bidi._mirrored_ui_layout = not Bidi._mirrored_ui_layout
+    end
+    -- Xtext default language and direction
+    if G_reader_settings:nilOrTrue("use_xtext") then
+        local xtext = require("libs/libkoreader-xtext")
+
+        -- Text direction should normally not follow ui mirroring
+        -- lang override (so that Arabic is still right aligned
+        -- when one wants the UI layout LTR). But allow it to
+        -- be independantly reversed (for testing UI mirroring
+        -- with english text right aligned).
+        if G_reader_settings:isTrue("reverse_lang_ui_text_direction") then
+            is_rtl = not is_rtl
+        end
+        xtext.setDefaultParaDirection(is_rtl)
+        Bidi._rtl_ui_text = is_rtl
+
+        -- Text language: this helps picking localized glyphs from the
+        -- font (eg. ideographs shaped differently for Japanese vs
+        -- Simplified Chinese vs Traditional Chinese).
+        -- Allow overriding xtext language rules from main UI language
+        -- (eg. English UI, with French line breaking rules)
+        local alt_lang = G_reader_settings:readSetting("xtext_alt_lang") or lang
+        if alt_lang then
+            xtext.setDefaultLang(alt_lang)
+        end
+    end
+    -- Optimise Bidi.wrap by aliasing it to the right wrapper
+    if Bidi._rtl_ui_text then
+        Bidi.wrap = Bidi.rtl
+    else
+        Bidi.wrap = Bidi.noop
+    end
+end
+
+
+-- Use this function in widgets to check if UI elements mirroring
+-- is to be done
+function Bidi.mirroredUILayout()
+    return Bidi._mirrored_ui_layout
+end
+
+-- This function might not be useful (RTL text is handled directly
+-- by TextWidget and TextBoxWidget)
+function Bidi.rtlUIText()
+    return Bidi._rtl_ui_text
+end
+
+-- Small helper to mirror gesture directions
+local mirrored_directions = {
+    east = "west",
+    west = "east",
+    northeast = "northwest",
+    northwest = "northeast",
+    southeast = "southwest",
+    southwest = "southeast",
+}
+
+function Bidi.mirrorDirectionIfMirroredUILayout(direction)
```

Comment:
Um, what? :-)

Reply:
You should see where they are used.
But ok, `flipDirectionIfMirroredUILayout()` to be less heavy ? :)

Reply:
Since flip is used elsewhere already it's probably better that way.